rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,132 @@
+* Product: @xinding33 || @klizhentas
+
+## What
+
+This RFD proposes a backend implementation and an audit log implementation using [PostgreSQL](https://www.postgresql.org/) as the underlying storage mechanism.
+
+## Why
+
+Currently, users looking to self-host Teleport in a High Availability (HA) deployment outside of AWS (where they could use DynamoDB) or GCP (where they could use Firestore) must be willing and capable to operate [etcd](https://etcd.io/), as that's the only HA-supporting, self-runnable backend currently supported by Teleport. In addition, they must handle shipping log files from all their auth servers into some long-term storage facility or forgo the use of the Teleport audit log, as there is no HA-capable audit log supported by Teleport that's self-hostable outside of said cloud environments.
+
+With the Postgres backend and audit log, users with infrastructure on Azure will be able to use the managed [Azure Database for PostgreSQL](https://azure.microsoft.com/en-us/products/postgresql/) service, with [Managed Identities](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/) allowing secretless authentication between Teleport and the database, and users running infrastructure on bare metal will have an option other than etcd for a HA deployment of Teleport, including a working audit log.
+
+## Details
+
+### Non-goals
+
+We don't intend for this backend and audit log to be extensible to completely different RDBMS (such as MariaDB or Microsoft SQL Server); support for other Postgres-compatible databases (such as CockroachDB) should be possible in the future, but is not in scope for this RFD.
+
+### General architecture of the backend
+
+The Teleport backend abstraction consists of a key-value store (where both key and value are opaque bytestrings), with a per-item optional Time-To-Live (TTL), and a stream of changes to the data. As Postgres provides a way to fetch changes to the data over time through [logical decoding](https://www.postgresql.org/docs/current/logicaldecoding.html) of the Write-Ahead Log, we can tailor the database schema to the exact needs of the backend operations.
+
+As such, the schema consists of a single `kv` table, defined as such:
+```pgsql
+CREATE TABLE kv (
+    key bytea NOT NULL,
+    value bytea NOT NULL,
+    expires timestamptz,
+    revision uuid NOT NULL,
+    CONSTRAINT kv_pkey PRIMARY KEY (key)
+);
+CREATE INDEX kv_expires_idx ON kv (expires) WHERE expires IS NOT NULL;
+```
+
+Each item in the backend is represented by a row in the `kv` table, with a `NULL` value for `expires` representing no expiration. To efficiently delete expired items, we keep a (partial) index on the expiration time. The `revision` column contains the revision for the upcoming opportunistic locking, and shall be set to a new unique value by the application code for every write (unfortunately, it's not a constraint that's expressible in SQL).
```

Comment:
> and shall be set to a new unique value by the application code for every write (unfortunately, it's not a constraint that's expressible in SQL).

Can we use triggers? Just generate UUID inside the trigger and write it to the `revision` field on update.

Reply:
I really don't want to get into maintaining plpgsql functions if we don't have to. 😅

A trigger would absolutely work in the single item case, but we have Plans™ for multi-item write operations in which it seems it could be useful for all items written at once to have the same revision - you could accomodate that by only regenerating the revision if it's different from the previous one for updates, or by generating one if the code leaves at a sentinel default for inserts, but it seems more trouble than it's worth, tbh.

Using some deterministic function of the txid would also work but it wouldn't be safe against a dump/restore of the database - although now that I'm thinking about it, maybe something based on time and txid would work?